### PR TITLE
Make Tools::getValue() mockable for unit test

### PR DESCRIPTION
**Pull request by thierrymarianne**
| Questions | Answers |
| --- | --- |
| Branch? | develop |
| Description? | Introduction of a test covering a method of `AdminTabsController` |
| Type? | improvement |
| Category? | CO |
| BC breaks? | no |
| Deprecations? | no |
| Fixed ticket? | N/A |
| How to test? | Run tests with PHPUnit |

### CODE REVIEW

**Comment by thierrymarianne** [comment]
@PrestaShop/prestashop-core-dev I would like to know what are your thoughts about this contrib and more particularly about changes made to `Tools::getValue` and `Tools::__construct` as they might enable us to add coverage for legacy classes.

I've introduced an instance of `\Symfony\Component\HttpFoundation\Request` as provider of HTTP parameters so that GET and POST values could be replaced easier in our tests via stubbing. I believe the use of an environment variable (''kernel.environment`) to switch from the use of superglobals to methods calls of Request could raise some debates. Would you proceed with alternate workarounds when dealing with static methods calls and testing?

File: classes/exception/PrestaShopException.php [hunk 1]
```diff
@@ -34,6 +34,10 @@ class PrestaShopExceptionCore extends Exception
      */
     public function displayMessage()
     {
+        if (getenv('kernel.environment') === 'test') {
+            throw $this;
```
Comment (mickaelandrieu) [inline classes/exception/PrestaShopException.php]: can we retrieve the error message display as a string instead ?

File: tests/Unit/controller/admin/AdminTabsControllerTest.php [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+use PrestaShop\PrestaShop\Tests\TestCase\UnitTestCase;
+use Phake;
+use Tools;
+use Tab;
+use DbQuery;
+use AdminTabsController;
+
+/**
+ * @package Prestashop\Prestashop\Tests\Unit\Controller\Admin
```
Comment (mickaelandrieu) [inline tests/Unit/controller/admin/AdminTabsControllerTest.php]: I'm not sure we use @package notation, even in tests

File: tests/phpunit-admin.xml [hunk 3]
```diff
@@ -0,0 +1,15 @@
+<phpunit bootstrap="bootstrap-admin.php">
+    <php>
+      <server name="kernel.environment" value="test" />
```
Comment (mickaelandrieu) [inline tests/phpunit-admin.xml]: ~~Why do you need to declare this server value ? isn't ``putenv`` and this declaration strictly equivalents ?

File: tests/phpunit.xml [hunk 4]
```diff
@@ -1,4 +1,9 @@
 <phpunit bootstrap="bootstrap.php">
+    <groups>
+        <exclude>
```
Comment (mickaelandrieu) [inline tests/phpunit.xml]: I guess we need to declare the launch of "admin" group tests in travis too :)

**Review by mickaelandrieu — COMMENTED** [review]
This is a very good move :+1:  Good job Thierry !

File: classes/exception/PrestaShopException.php (hunk 1 above)
Comment (thierrymarianne) [inline classes/exception/PrestaShopException.php]: Being re-thrown, the exception stack trace is be formatted by phpunit thanks to its internals formatter.

File: tests/phpunit-admin.xml (hunk 3 above)
Comment (thierrymarianne) [inline tests/phpunit-admin.xml]: It was setup in the xml config before I realize I needed it in the bootstrap as well. I'll remove it from the XML file.

File: tests/Unit/controller/admin/AdminTabsControllerTest.php (hunk 2 above)
Comment (thierrymarianne) [inline tests/Unit/controller/admin/AdminTabsControllerTest.php]: Well spotted. Thanks.

File: tests/phpunit.xml (hunk 4 above)
Comment (thierrymarianne) [inline tests/phpunit.xml]: I'm still figuring why the base suite currently fails. I will update travis config as soon as the pre-existing suite is passing again.

File: classes/exception/PrestaShopException.php (hunk 1 above)
Comment (mickaelandrieu) [inline classes/exception/PrestaShopException.php]: ok :)

File: tests/phpunit-admin.xml (hunk 3 above)
Comment (thierrymarianne) [inline tests/phpunit-admin.xml]: I've moved the environment variable configuration from the bootstrap script to the config file.

**Pull request merged**